### PR TITLE
Add commit hooks

### PR DESCRIPTION
**Pull request by jla524**
for #477

### CODE REVIEW

**Comment by jla524** [comment]
I couldn't get pyflakes to work properly. It doesn't ignore the lines with `# noqa`.

```
❯ pyflakes tinygrad/
tinygrad/__init__.py:1:1: 'tinygrad.tensor' imported but unused
tinygrad/__init__.py:1:1: 'tinygrad.nn' imported but unused
tinygrad/lazy.py:139:35: local variable 'ret' is assigned to but never used
```

**Comment by jla524** [comment]
`mypy` gives a lot of errors with the `--check-untyped-defs` flag. I'll try to fix these.

```
❯ mypy tinygrad/ --ignore-missing-imports --check-untyped-defs
tinygrad/shape/__init__.py:131: error: Argument 1 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:131: error: Argument 2 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:147: error: Argument 2 to "View" has incompatible type "List[int]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:171: error: Argument 2 to "View" has incompatible type "List[int]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:186: error: Argument 1 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:186: error: Argument 2 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:205: error: Argument 1 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:215: error: Argument 2 to "View" has incompatible type "List[int]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:224: error: Argument 1 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/shape/__init__.py:224: error: Argument 2 to "View" has incompatible type "List[Any]"; expected "Tuple[int, ...]"  [arg-type]
tinygrad/ast.py:122: error: Incompatible types in assignment (expression has type "Union[Literal[False], Any]", variable has type "List[Any]")  [assignment]
tinygrad/llops/ops_cpu.py:35: error: Cannot call function of unknown type  [operator]
tinygrad/llops/ops_cpu.py:36: error: Cannot call function of unknown type  [operator]
tinygrad/llops/ops_cpu.py:37: error: Cannot call function of unknown type  [operator]
tinygrad/llops/ops_cpu.py:38: error: Cannot call function of unknown type  [operator]
tinygrad/llops/ops_llvm.py:107: error: Item "None" of "Optional[Any]" has no attribute "target_data"  [union-attr]
tinygrad/llops/ops_llvm.py:115: error: Item "None" of "Optional[Any]" has no attribute "run"  [union-attr]
tinygrad/llops/ops_llvm.py:121: error: Item "None" of "Optional[Any]" has no attribute "emit_assembly"  [union-attr]
tinygrad/llops/ops_llvm.py:122: error: Item "None" of "Optional[Any]" has no attribute "add_module"  [union-attr]
tinygrad/llops/ops_llvm.py:123: error: Item "None" of "Optional[Any]" has no attribute "finalize_object"  [union-attr]
tinygrad/llops/ops_llvm.py:130: error: Item "None" of "Optional[Any]" has no attribute "get_function_address"  [union-attr]
tinygrad/llops/ops_llvm.py:141: error: Item "None" of "Optional[Any]" has no attribute "remove_module"  [union-attr]
tinygrad/llops/ops_llvm.py:313: error: Cannot call function of unknown type  [operator]
tinygrad/llops/ops_llvm.py:318: error: Cannot call function of unknown type  [operator]
tinygrad/tensor.py:165: error: "Tensor" has no attribute "slice"  [attr-defined]
tinygrad/tensor.py:171: error: Incompatible types in assignment (expression has type "List[Any]", variable has type "Tuple[Any, ...]")  [assignment]
tinygrad/tensor.py:176: error: "Type[Tensor]" has no attribute "__iadd__"  [attr-defined]
tinygrad/tensor.py:183: error: "Tensor" has no attribute "slice"  [attr-defined]
tinygrad/tensor.py:222: error: "Type[Tensor]" has no attribute "_sum"  [attr-defined]
tinygrad/tensor.py:223: error: "Type[Tensor]" has no attribute "_max"  [attr-defined]
tinygrad/tensor.py:247: error: Unsupported left operand type for * ("Tensor")  [operator]
tinygrad/tensor.py:258: error: "Tensor" has no attribute "_conv2d"; maybe "conv2d"?  [attr-defined]
tinygrad/tensor.py:263: error: Unsupported operand types for - ("float" and "Tensor")  [operator]
tinygrad/tensor.py:265: error: Unsupported left operand type for * ("Tensor")  [operator]
tinygrad/tensor.py:267: error: "Tensor" has no attribute "relu"; maybe "relu6" or "elu"?  [attr-defined]
tinygrad/tensor.py:273: error: "Tensor" has no attribute "relu"; maybe "relu6" or "elu"?  [attr-defined]
tinygrad/tensor.py:273: error: "Tensor" has no attribute "exp"  [attr-defined]
tinygrad/tensor.py:276: error: "Tensor" has no attribute "relu"; maybe "relu6" or "elu"?  [attr-defined]
tinygrad/tensor.py:276: error: "int" has no attribute "relu"  [attr-defined]
tinygrad/tensor.py:276: error: Unsupported operand types for - ("Tensor" and "int")  [operator]
tinygrad/tensor.py:277: error: "int" has no attribute "relu6"  [attr-defined]
tinygrad/tensor.py:277: error: Unsupported operand types for + ("Tensor" and "int")  [operator]
tinygrad/tensor.py:278: error: "float" has no attribute "sigmoid"  [attr-defined]
tinygrad/tensor.py:278: error: Unsupported operand types for * ("float" and "Tensor")  [operator]
tinygrad/tensor.py:279: error: Unsupported operand types for * ("float" and "Tensor")  [operator]
tinygrad/tensor.py:279: error: "float" has no attribute "tanh"  [attr-defined]
tinygrad/tensor.py:279: error: Unsupported operand types for * ("Tensor" and "float")  [operator]
tinygrad/tensor.py:280: error: "float" has no attribute "sigmoid"  [attr-defined]
tinygrad/tensor.py:280: error: Unsupported operand types for * ("Tensor" and "float")  [operator]
tinygrad/tensor.py:281: error: "Tensor" has no attribute "relu"; maybe "relu6" or "elu"?  [attr-defined]
tinygrad/tensor.py:296: error: "Type[Tensor]" has no attribute "_add"  [attr-defined]
tinygrad/tensor.py:297: error: "Type[Tensor]" has no attribute "_sub"  [attr-defined]
tinygrad/tensor.py:298: error: "Type[Tensor]" has no attribute "_mul"  [attr-defined]
tinygrad/tensor.py:299: error: "Type[Tensor]" has no attribute "_pow"  [attr-defined]
tinygrad/tensor.py:300: error: "Tensor" has no attribute "reciprocal"  [attr-defined]
tinygrad/tensor.py:305: error: "Tensor" has no attribute "_reshape"; maybe "reshape" or "shape"?  [attr-defined]
tinygrad/tensor.py:306: error: "Tensor" has no attribute "_expand"; maybe "expand"?  [attr-defined]
tinygrad/tensor.py:307: error: "Tensor" has no attribute "_permute"; maybe "permute"?  [attr-defined]
tinygrad/llops/ops_gpu.py:55: error: Need type annotation for "devices"  [var-annotated]
tinygrad/llops/ops_gpu.py:90: error: Item "None" of "Optional[Any]" has no attribute "finish"  [union-attr]
tinygrad/llops/ops_gpu.py:356: error: Invalid index type "Union[UnaryOps, BinaryOps, ReduceOps, MovementOps, ProcessingOps, LoadOps]" for "Dict[ReduceOps, str]"; expected type "ReduceOps"  [index]
tinygrad/llops/ops_gpu.py:376: error: Item "None" of "Optional[LazyOp]" has no attribute "op"  [union-attr]
tinygrad/llops/ops_gpu.py:378: error: Item "None" of "Optional[LazyOp]" has no attribute "op"  [union-attr]
tinygrad/llops/ops_gpu.py:420: error: Incompatible types in assignment (expression has type "Union[CLImage, CLBuffer]", variable has type "Optional[CLBuffer]")  [assignment]
tinygrad/llops/ops_gpu.py:422: error: Item "None" of "Optional[CLBuffer]" has no attribute "cl"  [union-attr]
tinygrad/llops/ops_gpu.py:424: error: Item "None" of "Optional[CLBuffer]" has no attribute "cl"  [union-attr]
Found 66 errors in 6 files (checked 16 source files)
```

File: tinygrad/shape/__init__.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def to_shape_strides(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> List[Tup
   return ret
 
 class View:
-  def __init__(self, shape:Tuple[int, ...], strides:Tuple[int, ...], offset:int=0):
+  def __init__(self, shape:Union[Tuple[int, ...],List[Any]], strides:Union[Tuple[int, ...],List[Any]], offset:int=0):
```
Comment (geohot) [inline tinygrad/shape/__init__.py:24]: This should be List or Tuple somehow, at least it should be List[int]

**Comment by geohot** [comment]
Merged anyway, thanks for the hooks!

**Comment by geohot** [comment]
Want to bring back #415? It broke stable diffusion on my computer, wonder if there's a way to write a fast test that stable diffusion runs without needing to load or run the model?

**Comment by jla524** [comment]
> Want to bring back #415? It broke stable diffusion on my computer, wonder if there's a way to write a fast test that stable diffusion runs without needing to load or run the model?

I couldn't figure out how to get stable diffusion to work without the "elementwise_affine on channels" hack. But I'll try and bring it back when I'm more familiar with transformers.

**Pull request merged**